opennav_following_bt/include/opennav_following_bt/follow_object.hpp

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2024 Open Navigation LLC
+// Copyright (c) 2024 Alberto J. Tudela Roldán
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef OPENNAV_FOLLOWING_BT__FOLLOW_OBJECT_HPP_
+#define OPENNAV_FOLLOWING_BT__FOLLOW_OBJECT_HPP_
+
+#include <string>
+#include <memory>
+#include <vector>
+
+#include "opennav_following_msgs/action/follow_object.hpp"
+#include "geometry_msgs/msg/pose_stamped.hpp"
+#include "nav2_behavior_tree/bt_action_node.hpp"
+
+namespace opennav_following_bt
+{
+
+/**
+ * @brief nav2_behavior_tree::BtActionNode class that wraps opennav_following_msgs/FollowObject
+ */
+class FollowObjectAction
+  : public nav2_behavior_tree::BtActionNode<
+    opennav_following_msgs::action::FollowObject>
+{
+  using Action = opennav_following_msgs::action::FollowObject;
+  using ActionResult = Action::Result;
+
+public:
+  /**
+   * @brief A constructor for opennav_following_bt::FollowObjectAction
+   * @param xml_tag_name Name for the XML tag for this node
+   * @param action_name Action name this node creates a client for
+   * @param conf BT node configuration
+   */
+  FollowObjectAction(
+    const std::string & xml_tag_name,
+    const std::string & action_name,
+    const BT::NodeConfiguration & conf);
+
+  /**
+   * @brief Function to perform some user-defined operation on tick
+   */
+  void on_tick() override;
+
+  /**
+   * @brief Function to perform some user-defined operation upon successful completion of the action
+   */
+  BT::NodeStatus on_success() override;
+
+  /**
+   * @brief Function to perform some user-defined operation upon abortion of the action
+   */
+  BT::NodeStatus on_aborted() override;
+
+  /**
+   * @brief Function to perform some user-defined operation upon cancellation of the action
+   */
+  BT::NodeStatus on_cancelled() override;
+
+  /**
+   * \brief Override required by the a BT action. Cancel the action and set the path output
+   */
+  void halt() override;
+
+  /**
+   * @brief Creates list of BT ports
+   * @return BT::PortsList Containing basic ports along with node-specific ports
+   */
+  static BT::PortsList providedPorts()
+  {
+    return providedBasicPorts(
+      {
+        BT::InputPort<geometry_msgs::msg::PoseStamped>(
```

Comment:
Wouldn't this be continually moving? Shouldn't the server obtain the pose to follow from a topic or are you consistently preempting this action with new request poses as a detector runs in the BT?

Reply:
This is a part of our use case. We have a people detection system that publishes a person_msg array (each person has a name and a pose). The ```GetPersonPose``` filters / extracts the selected person we and outputs the pose to the ```FollowObject``` BT. This pose is used as the object_pose in the request.

On the other hand, ```GetPersonPose``` publishes the ```/detected_dynamic_pose``` with the topic that the FollowObject listens to. 

Both BTs are ticked once, but they are kept in the running state.

![BT](https://github.com/user-attachments/assets/b619da28-6aeb-4897-a6e3-ba38e8d1e0ab)

Reply:
So this is being regularly preempted? Wouldn't it be better for the server to listen to the topic instead? 

I mostly say this because I'm weary about preempting actions at 10-20-50-100hz that's alot going on that I have not used the Actions API to do before. On the other hand, the subscriptions its very normal to have high rate data and we can cleanly just drop old detections that aren't important to us anymore. Actions will try to process each one coming in and then relies on the queuing method of the actions QoS and SimpleActionServer instead of just using a subscription queue. 

I suppose it could work this way, but its a bit off the beaten path.

Reply:
The BT calls the action once with the first pose. The action starts executing and the BT remains in a running state until the action is finished. There is no preemption or cancellation. I know that having the first pose as the goal request of the action, and then publishing (and continuing to publish) the same pose and the next ones as a topic, is not the best solution.

Do you think iit would be best to remove the ```object_pose``` completely from the goal request and just have a topic that listens for pose changes? The topic could be set as a parameter.

Reply:
Ah yeah, we should do one or the other: have it given to the action in the request and preempted or have it always provided from a topic. 

> Do you think iit would be best to remove the object_pose completely from the goal request and just have a topic that listens for pose changes? The topic could be set as a parameter.

I think that would be best :-)